### PR TITLE
ENH: Add optional `anaconda_nightly_upload_organization` argument

### LINKED ISSUE #46: Allow a new inputs parameter named `anaconda_nightly_upload_url`
Hi Scientific Python Team, 

is it possible to add the optional input `anaconda_nightly_upload_url`.

Instead of creating a different fork, I would like to reuse this action for another anaconda repo (DIPY nightly wheels)

it would be great to have also the optional input label (manage anaconda label).

Let me know what you think, if it is possible to do that here or if I should fork and maintain an alternative version.

thank you for the feedback.

### PR DESCRIPTION
**Pull request by skoudoro**
This is a follow up and fixes #46

This PR goal is to allow other Scientific Python projects to use this github action.

Please, let me know what you think, and I will be happy to rename the variables if needed.

Thanks  

**Squash and merge commit**

```
* Add optional `anaconda_nightly_upload_organization` and
  `anaconda_nightly_upload_labels` arguments to allow for packages
  to be uploaded to custom registires with the GitHub Action.
* Add docs to README providing an example of how to upload to an
  alternative registry.
```

### CODE REVIEW

File: action.yml
Comment (matthewfeickert) [inline action.yml:24]: Why is the label needed? This is a change in current behavior.

**Review by matthewfeickert — COMMENTED** [review]
I'm on a plane and so reviewing on the GitHub mobile app (which isn't the best, sorry) but can you elaborate on why labels are needed now? They shouldn't be AFAIK.

File: action.yml
Comment (skoudoro) [reply action.yml:24]: By default, the label is main when you upload an artifact in anaconda

image

in our case (https://anaconda.org/dipy/dipy), the organization name is not explicit. I would like to avoid random people downloading nightly wheels and believing that is it a release. So, we tag them as dev instead of main (see the difference between the sreenshot above and below

image

one artifact can have one or multiple labels. I am fine for alternative solution to avoid confusion with users (since I can relabel all the old artifacts)

Documentation on anaconda label: https://docs.anaconda.com/free/anacondaorg/user-guide/work-with-labels/

**Comment by matthewfeickert** [comment]
> I'm on a plane and so reviewing on the GitHub mobile app (which isn't the best, sorry) but can you elaborate on why labels are needed now? They shouldn't be AFAIK.

Oh I guess this isn't a change in behavior given the current channel uploads

image 

My bad. Can you document this optional variable though in your example?

File: README.md
Comment (stefanv) [inline README.md:50]:
```suggestion
define the `anaconda_nightly_upload_url` variable. Furthermore,
```

File: README.md
Comment (stefanv) [inline README.md:51]:
```suggestion
you can add labels for organizing your artifacts using `anaconda_nightly_upload_labels`
```

File: README.md
Comment (stefanv) [inline README.md:64]:
```suggestion
        anaconda_nightly_upload_labels: main
```

This is the default, so how about using a different label in the example?

File: action.yml
Comment (stefanv) [inline action.yml:18]:
```suggestion
    description: 'URL of the index to upload the wheels to'
```

File: action.yml
Comment (stefanv) [inline action.yml:20]:
```suggestion
    default: scientific-python-nightly-wheels
```

I don't know how the action works, but this is confusing since it isn't a URL.
If it can be anything other than a URL, that should either be documented or the field be renamed.

File: action.yml
Comment (stefanv) [inline action.yml:24]:
```suggestion
    default: [main]
```

If multiple labels are accepted, then this would make it clearer.

**Review by stefanv — COMMENTED** [review]
Thanks @skoudoro! It looks good to me; I left some feedback.

File: README.md
Comment (skoudoro) [reply README.md:64]: agree, I will update that

File: action.yml
Comment (skoudoro) [reply action.yml:18]: goood catch !

File: action.yml
Comment (skoudoro) [reply action.yml:20]: So the variable name is wrong! it should be `anaconda_nightly_upload_organization`

**Review by stefanv — APPROVED** [review]
Looks good to me; I'll let @matthewfeickert do the final review and merge.

**Comment by matthewfeickert** [comment]
> Looks good to me; I'll let @matthewfeickert do the final review and merge.

Thanks all! 👍 I'm running in airports so if I have wifi on next flight I'll knock this final look over out and if not then I'll get to it tomorrow afternoon Europe time.

File: action.yml
Comment (matthewfeickert) [inline action.yml:24]: @skoudoro Hey sorry to have one final question here (now that I've arrived), but can you point me to a reference on how you know that `anaconda upload --label` will accept a list of labels (as opposed to repeated calls of `--label` like how you would have to do with `--channel` with `micromamba install`? `micromamba install --channel main --channel conda-forge numpy`)?

The `anaconda-client` docs are not good, so it is possible I'm missing something. Have you tested this with any uploads to your registry yet?

**Review by matthewfeickert — CHANGES_REQUESTED** [review]

File: cmd.sh
Comment (bsipocz) [inline cmd.sh:17]: would this break the current behaviour for the core libraries uploading to the default location?

File: action.yml
Comment (skoudoro) [reply action.yml:24]: > can you point me to a reference on how you know that anaconda upload --label will accept a list of labels

As you can see below, I just used the help of the command line (`anaconda upload --help`.

image

There is also the following documentation : https://docs.anaconda.com/anaconda-repository/user-guide/tutorials/#using-labels-in-the-development-cycle

>  Have you tested this with any uploads to your registry yet?

Yes,  I tried with some artifacts on our repository. I use the following command manually:
`anaconda upload --force --user dipy --label dev --label main  *1.6.0*.whl`

File: cmd.sh
Comment (skoudoro) [reply cmd.sh:17]: No, because I added the default location in this line: https://github.com/scientific-python/upload-nightly-action/pull/47/files#diff-1243c5424efaaa19bd8e813c5e6f6da46316e63761421b3e5f5c8ced9a36e6b6R20

However, I believe that:

1. this parameter should be required instead of optional  
2. After merging, we should ask the core libraries to add and update this variable
3. Then, when it is done, create a new PR here to remove this default location and let it empty.

Thank you for your review @bsipocz!

File: action.yml
Comment (skoudoro) [reply action.yml:24]: Not sure how to handle the list with the docker. I am looking forward for your feedback

File: cmd.sh
Comment (matthewfeickert) [reply cmd.sh:17]: @skoudoro While I'm not against the idea of being explicit about the target channel (I think that's a good idea in general to guard against default changes), I think the original intent for the GitHub Action was to make uploading as easy as possible for the core packages with the idea that they need to configure things as minimally as possible.

I think that this could probably be discussed more broadly outside of this PR though, so maybe you can open up an Issue from your comment to discuss what the target audience and scope of the action configuration should be?

**Review by matthewfeickert — CHANGES_REQUESTED** [review]

File: action.yml
Comment (skoudoro) [reply action.yml]: Yes, agree. 

I just read more about that and make sense

Comment (skoudoro) [reply action.yml]: yes, I just updated the code. This is not a list, sorry for the confusion.

>  how you're using labels now and why you want to have multiple labels?

- We use only the label dev for now
- in our case (https://anaconda.org/dipy/dipy), the organization name is not explicit for nightly wheels. I would like to avoid random people downloading nightly wheels and believing that is it a release. So, we tag them as dev instead of main. it force user to be explicit and conscious of what they are trying to get. At the same time, it is not an issue for downstream packages.
- we do not use multiple labels as the same time. the line above was just an example to show you how it can be done.

**Comment by matthewfeickert** [comment]
(Sorry, been traveling for work to CERN and so haven't gotten back to reviewing this. Will try this weekend.)

**Comment by skoudoro** [comment]
No worries @matthewfeickert , take your times

**Comment by skoudoro** [comment]
To help you for the review I use this action here: https://github.com/dipy/dipy/blob/master/.github/workflows/nightly.yml#L136-L156

and an example of the action result: https://github.com/dipy/dipy/actions/runs/7148367041/job/19471063289#step:4:376

**Comment by bsipocz** [comment]
ping @matthewfeickert, this would be good to have in sooner rather than later to help various downstream wheel production

**Comment by matthewfeickert** [comment]
> ping @matthewfeickert, this would be good to have in sooner rather than later to help various downstream wheel production

Yes, I'm aware but c.f. again https://github.com/scientific-python/upload-nightly-action/issues/50#issuecomment-1872476726 for why I've been preoccupied. I'll have a stable computer and get to this before end of week.

**Comment by bsipocz** [comment]
> c.f. again https://github.com/scientific-python/upload-nightly-action/issues/50#issuecomment-1872476726 for why I've been preoccupied

Hah, fair enough, sorry for rushing you. I was also traveling, so I didn't really internalize all the comments from the past few weeks.

File: action.yml
Comment (matthewfeickert) [inline action.yml:22]:
```suggestion
    description: 'List of labels assigned to the uploaded artifacts'
```

As this could be a list it should be stated in the description.

File: cmd.sh
Comment (matthewfeickert) [inline cmd.sh:24]:
```suggestion
  echo "ANACONDA_ORG is empty, exiting..."
```

I know that there is an extra accidental space above, but that's a typo we don't need to recreate. (this is super nitpicky, sorry, which is why I'm adding it as a suggested fix so you don't have to do anything)

File: cmd.sh
Comment (matthewfeickert) [inline cmd.sh:52]:
```suggestion
# if the ANACONDA_LABELS is empty, exit with status -1
# as this should be set in action.yml or by the user
# and it is better to fail on this to sigal a problem.
if [ -z "${ANACONDA_LABELS}" ]; then
  echo "ANACONDA_LABELS is empty, exiting..."
  exit -1
fi

# convert newlines to commas for parsing
# and ensure that there is no trailing comma
ANACONDA_LABELS="$(echo "$ANACONDA_LABELS" | tr '\n' ',' | sed 's/,$//')"

IFS=',' read -ra LABELS <<< "${ANACONDA_LABELS}"

LABEL_ARGS=""
for label in "${LABELS[@]}"; do
  LABEL_ARGS+="--label ${label} "
done
```

Okay, here I am going to suggest a total rewrite so let me explain why. At the moment the approach in the PR is assuming that the list of possible labels is space seperated. This is not standard across GitHub Actions, which in general assume that you're providing a newline seperatd or comma seperated list (c.f. https://github.com/docker/build-push-action?tab=readme-ov-file#inputs as an example).

So if we agree that we should be expecting newline seperated or comma seperated inputs then we can just treat all input the same by:
* First, converting all input into a comma separated string
* Then, create an array of the comma seperated labels
* Finally, parse that array into a single string that represents all the label arguments to be included (you were already doing this part)

This approach doesn't care about the number of labels used, and so the logic is very simple to impliment (and read in the future).

Examples of possible `ANACONDA_LABELS` inputs and the resulting `LABEL_ARGS` with the suggested changes:

* `ANACONDA_LABELS="main"`
```
LABEL_ARGS: --label main
```

* `ANACONDA_LABELS="main,dev"`
```
LABEL_ARGS: --label main --label dev
```

*
```
ANACONDA_LABELS="main
dev"
```

```
LABEL_ARGS: --label main --label dev
```

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
@skoudoro Okay, sorry for a ridiculously long delay (😞) in this review from me (I've been without a working computer for multiple weeks). So I think this is overall good and the right approach, but the one important change is that this approach currenlty assume a **space seperated** list of labels, which is not a common assumption in the world of GitHub Actions where newline seperted or comma seperated lists are generally expected.

The suggested changes I've added in this review address that and change the logic to be able to handle newline or comma seperated lists of lables. I've tried to step through things to make the reasoning/motivation clear, but if you have any questions please feel free to ask (I promise I will be much more responsive now!).

File: cmd.sh
Comment (matthewfeickert) [reply cmd.sh:52]: If you also did want to allow for space deliminated lists of labels then you could just throw in a 

```bash
ANACONDA_LABELS="$(echo "$ANACONDA_LABELS" | tr ' ' ',' | sed 's/,$//')"
```

to the suggestions I made as well.

**Comment by skoudoro** [comment]
Hi @matthewfeickert,

Thank you for the review and suggestions!  I will look at it this evening or tomorrow morning and come back to you.

Thanks again

**Comment by bsipocz** [comment]
gentle ping @skoudoro, as there aren't many outstanding review points here and I would really like to see this merged in  (I know is unfair to rush, so my apologies, as the PR was on our side for too long, too).

**Comment by stefanv** [comment]
In general, I would urge reviewers who want to see major refactors when a PR is close to its conclusion to push changes (when confident) or make PRs (of they want the contributor to review). This helps to keep things moving, and preserves contributor motivation.

**Comment by skoudoro** [comment]
Sorry, I was traveling and then sick. I will get back to it by the end of the week (trying to catch up with everything). Apologize for that

**Comment by bsipocz** [comment]
> In general, I would urge reviewers who want to see major refactors when a PR is close to its conclusion to push changes (when confident) or make PRs (of they want the contributor to review). This helps to keep things moving, and preserves contributor motivation.

I'm happy to do this (and am agree and tend to do this in places where I'm the primary maintainer and only have bikeshed comments), but feel that overtaking a PR where the contributor said they would come back to address another reviewer's comments can perceived as aggressive rather than helpful and I definitely would like to avoid that.

This PR was stuck on the reviewers' side for long enough that there is nothing to apologize for @skoudoro!

**Comment by matthewfeickert** [comment]
> In general, I would urge reviewers who want to see major refactors when a PR is close to its conclusion to push changes (when confident) or make PRs (of they want the contributor to review). This helps to keep things moving, and preserves contributor motivation.

@stefanv I think I share @bsipocz's feelings here, in that I am not a big fan of reviewrs pushing changes to code unless explictly asked to, which is why I always use GitHub's suggest changes tool so that if people agree with them they can incoperate as many as they wish into the PR with a simple button click and don't have to do any additional work themselves.

If pushing changes is preferred here I'm fine with that, and I will just accept my own proposed changes through the UI as a single commit. I'm happy to wait though to let @skoudoro have time to look at the proposed changes later this week.

**Comment by stefanv** [comment]
Using the inline commenting system is indeed friendly in that regard, but when larger pieces of code needs to be re-written it becomes impractical and a PR to the contributor branch does the same.

We should, however, not be shy to help one another shoulder the load; I've never heard of a maintainer complain because other maintainers helped them too much. We all have long TODO lists that will never be completed in this lifetime!

**Comment by bsipocz** [comment]
> We should, however, not be shy to help one another shoulder the load; I've never heard of a maintainer complain because other maintainers helped them too much.

fair enough, my experience with people being annoyed about PRs being taken over was PRs from generic (scientist or student) contributors rather than PRs from maintainers of other packages

**Comment by matthewfeickert** [comment]
> We should, however, not be shy to help one another shoulder the load; I've never heard of a maintainer complain because other maintainers helped them too much. We all have long TODO lists that will never be completed in this lifetime!

Okay, given this I'm going to accept my suggested changes as a single commit. @skoudoro to be explicit here, _please_ take this as just a suggestion and feel free to revert or make any futher changes that you think would improve things. I'm happy to iterate with you further if needed! (though in full disclousure I have a job interview next week that is going to take me offline for most of it :grimacing: ).

**Comment by stefanv** [comment]
Thanks @matthewfeickert, and good luck with the interview!

**Comment by stefanv** [comment]
Test failure :eyes: 

> ANACONDA_TOKEN is empty , exiting...

**Comment by matthewfeickert** [comment]
> Test failure 👀
> 
> > ANACONDA_TOKEN is empty , exiting...

That's expected. c.f. https://github.com/scientific-python/upload-nightly-action/pull/32#issuecomment-1678400638 for a not great summary, but basically there isn't a great solution at the moment to run the tests on PRs from froks that wouldn't potentially open up a security threat.

I'm not saying it is impossible to fix, but it would require a redesign of the CI.

**Comment by stefanv** [comment]
That makes sense; I run into this all the time.

**Review by matthewfeickert — APPROVED** [review]

**Review by bsipocz — APPROVED** [review]

**Comment by matthewfeickert** [comment]
@skoudoro Can you verify that if you use 

```yaml
jobs:
  steps:
    ...
    - name: Upload wheel
      uses: scientific-python/upload-nightly-action@3eb3a42b50671237cace9be2d18a3e4b3845d3c4f  # currently main
      with:
        artifacts_path: dist
        anaconda_nightly_upload_organization: my-alternative-organization
        anaconda_nightly_upload_token: ${{secrets.UPLOAD_TOKEN}}
        anaconda_nightly_upload_labels: dev
```

that things work for you to upload to your alternative registry? If so, I'll tag this commit as `v0.3.0`.

**Comment by bsipocz** [comment]
cc @nabobalis if you want to try this out for openastronomy/sunpy

**Comment by skoudoro** [comment]
Thank you all for pushing that forward.

@matthewfeickert, I just pushed the changes on https://github.com/dipy/dipy/commit/66d08be02b0abe87b317b03c8977e82c24f284dd. This is running

**Comment by nabobalis** [comment]
> cc @nabobalis if you want to try this out for openastronomy/sunpy

Let me try.

**Comment by skoudoro** [comment]
> @skoudoro Can you verify that if you use
> 
> ```yaml
> jobs:
>   steps:
>     ...
>     - name: Upload wheel
>       uses: scientific-python/upload-nightly-action@3eb3a42b50671237cace9be2d18a3e4b3845d3c4f  # currently main
>       with:
>         artifacts_path: dist
>         anaconda_nightly_upload_organization: my-alternative-organization
>         anaconda_nightly_upload_token: ${{secrets.UPLOAD_TOKEN}}
>         anaconda_nightly_upload_labels: dev
> ```
> 
> that things work for you to upload to your alternative registry? If so, I'll tag [3eb3a42](https://github.com/scientific-python/upload-nightly-action/commit/3eb3a42b50671237cace9be2d18a3e4b3845d3c4) as `v0.3.0`.

ok, it works !  Thanks @matthewfeickert ! Make sure to put the right commit number @nabobalis.  The little code above was confusing (wrong commit number)

**Comment by matthewfeickert** [comment]
> The little code above was confusing (wrong commit number)

@skoudoro what's wrong with 3eb3a42b50671237cace9be2d18a3e4b3845d3c4 ?

**Comment by matthewfeickert** [comment]
Okay, great as things seem to be working for [`dipy`'s uploads](https://anaconda.org/dipy/dipy/files)

[image](https://anaconda.org/dipy/dipy/files)

I'll go ahead and make a new tag and release.

Thank you for the PR and for verifying this works, @skoudoro!

**Pull request merged**
